### PR TITLE
ci: publish the minimal SDK in a continuous release

### PR DESCRIPTION
**Pull request by dscho**
Currently, when asking the [`setup-git-for-windows-sdk` Action](https://github.com/git-for-windows/setup-git-for-windows-sdk/) for the minimal SDK, it first looks at the latest successful `ci-artifacts` run to determine the commit, and then performs a shallow, partial & sparse checkout of the `git-sdk-64` repository at that commit.

This is relatively slow, taking typically something around 25 seconds for the actual clone, compared to typically around 3 seconds to restore it from a cached version. Therefore, the Action tries to cache it, even if it sometimes adds quite an overhead (for example, [here](https://github.com/git-for-windows/git/actions/runs/10838555769/job/30077050246#step:3:1) _something_ spent 2 minutes after the cache was saved before the workflow was allowed to continue).

Not only is this taking quite some time, diminishing the developer experience, but it is also unnecessary, seeing as the `ci-artifacts` run already provides the minimal SDK in a ready-to-use form and uploads it as a workflow artifact.

So why not use that workflow artifact? A couple of reasons:

- It would appear as if workflow artifacts are somehow compressed as zip archives on the fly when trying to download them manually,
- While not possible at the time `setup-for-windows-sdk` was designed, [it now seems possible](https://github.com/actions/download-artifact?tab=readme-ov-file#download-artifacts-from-other-workflow-runs-or-repositories) to use `actions/download-artifact` to download from other repositories, but:
- Workflow run artifacts obey a limited retention policy, which is undesirable here, and finally:
- You can only download workflow artifacts when authenticated, while clones can be performed anonymously; There are likely use cases where the minimal SDK is needed outside of GitHub Actions (for example in other CI systems).

I once looked into publishing the minimal SDK continuously, first considering [GitHub Packages](https://github.com/features/packages). But they have [rather strict retention rules](https://docs.github.com/en/packages/learn-github-packages/deleting-and-restoring-a-package#package-deletion-and-restoration-support-on-github) and we definitely do not need to retain anything but the latest working minimal-sdk artifact. (And I don't want to waste resources unnecessarily, even if someone else pays for them.)

Then I had looked into uploading that artifact to Azure Blobs, but [had to pull the plug in one big hurry](https://github.com/git-for-windows/git-sdk-64/commit/e51b79eef9230f4eb4a89719030b26206a33f947) because it threatened to blow my Azure budget.

So eventually I gave up on that plan and instead fell back to the "partial, shallow, sparse clone & then cache" strategy, which is in place to this day.

However, people who are more creative than I am figured out ways to publish artifacts continuously. I just had not found those solutions yet at the time. In the MSYS2 project, for example, they have a `srcinfo` artifact that is updated continuously, and initially they used [`eine/tip`](https://github.com/eine/tip?tab=readme-ov-file) to publish it in a GitHub release that is kept up to date by replacing the assets as new revisions come in. Nowadays they use the same strategy, albeit manually via [`gh release upload --clobber`](https://cli.github.com/manual/gh_release_upload#options).

This here PR uses a slight variation of the same strategy: Since `--clobber` basically [deletes](https://github.com/cli/cli/blob/v2.56.0/pkg/cmd/release/shared/upload.go#L135) the asset before [uploading a new version](https://github.com/cli/cli/blob/v2.56.0/pkg/cmd/release/shared/upload.go#L141), there is a time window where the asset is not available. Since the `.tar.gz` file is around 95MB, uploading it can take a while, in particular given the vagaries of any network operation. For that reason, in this PR the upload is done to a temporary name, then the old asset is deleted, and finally the new asset is renamed to the old name. This does not avoid the window of a missing asset altogether, but at least it makes it a lot smaller.

Finally, I took this opportunity to generate and publish `.zip` and `.7z.exe` variants, the latter being a self-extracting 7-Zip archive that does not require any special tools to extract. Granted, since [a native BSD tar is available in `C:\Windows\system32\tar.exe` on pretty much any non-ancient Windows version](https://devblogs.microsoft.com/commandline/tar-and-curl-come-to-windows/), and that `tar.exe` can extract both `.tar.gz` and `.zip` archives without problems, that seems to be lesser of an issue. But still, there's things like [Windows Nano Server](https://hub.docker.com/r/microsoft/windows-nanoserver) that may not have that BSD tar.

tl;dr with this PR, the `ci-artifacts` workflow runs not only verify that the minimal SDK is able to compile a working `git.exe`, but then also publishes the minimal SDK in various variants in a `git-sdk-64` GitHub release called `ci-artifacts` whose tag will point to the corresponding commit.

### CODE REVIEW

**Comment by dscho** [comment]
For the record, [here](https://github.com/dscho/git-sdk-64/releases/tag/ci-artifacts) is the `ci-artifacts` release in my personal fork that was done with the code presented here (plus a small commit on top to convince the workflow to run and publish in a fork).

File: .github/workflows/ci-artifacts.yml [hunk 1]
```diff
@@ -28,31 +32,22 @@ jobs:
...
       - name: compress artifact
         shell: bash
-        run: (cd minimal-sdk && tar cvf - * .[0-9A-Za-z]*) | gzip -1 >git-sdk-64-minimal.tar.gz
+        run: (cd minimal-sdk && tar cvf - * .[0-9A-Za-z]*) | gzip -1 >git-sdk-x86_64-minimal.tar.gz
       - name: upload minimal-sdk artifact
         uses: actions/upload-artifact@v4
         with:
           name: minimal-sdk
-          path: git-sdk-64-minimal.tar.gz
+          path: git-sdk-x86_64-minimal.tar.gz
```
Comment (dennisameling) [inline .github/workflows/ci-artifacts.yml:43]: Thanks for the rename here as well!

File: .github/workflows/ci-artifacts.yml [hunk 2]
```diff
@@ -64,6 +59,24 @@ jobs:
...
+        if: github.event_name == 'push' && github.ref == 'refs/heads/main'
+        shell: bash
+        run: |
+          for path in mingw64/bin/7z.exe mingw64/bin/7z.dll mingw64/lib/7zip/7zCon.sfx
+          do
+            git --git-dir=git-sdk-64.git show HEAD:$path >${path##*/}
+          done &&
+          mkdir minimal-sdk-extra &&
+          (cd minimal-sdk && ../7z.exe a -mx9 ../minimal-sdk-extra/git-sdk-x86_64-minimal.zip * .?*) &&
+          (cd minimal-sdk && ../7z.exe a -t7z -m0=lzma -mqs -mlc=8 -mx=9 -md=256M -mfb=273 -ms=256M -sfx../7zCon.sfx \
```
Comment (dennisameling) [inline .github/workflows/ci-artifacts.yml:72]: IIRC (I believe I saw it some time last year when testing some pipelines on a self-hosted runner), 7-Zip can't use multiple cores for creating some types of archives. Not sure if that applies here as well, but I wonder if running these commands in parallel would speed up [the 2m+ runtime](https://github.com/dscho/git-sdk-64/actions/runs/10851317340/job/30114789179) here. GitHub recently [upgraded](https://github.blog/news-insights/product-news/github-hosted-runners-double-the-power-for-open-source/) OSS runners to have 4 cores, so that might help us here! WDYT?

File: .github/workflows/ci-artifacts.yml [hunk 3]
```diff
@@ -164,3 +176,91 @@ jobs:
...
+                throw e;
+              }
+            })()
+
+            const fs = require('fs')
+            for (const fileName of [
+              'git-sdk-x86_64-minimal.tar.gz',
+              'git-sdk-x86_64-minimal.zip',
+              'git-sdk-x86_64-minimal.7z.exe',
+            ]) {
```
Comment (dennisameling) [inline .github/workflows/ci-artifacts.yml:228]: [6 seconds](https://github.com/dscho/git-sdk-64/actions/runs/10851317340/job/30115544965#step:4:81) is already really fast, but I wonder if we want to handle these uploads in parallel to potentially speed things up further. But it's a really minor nit, so please feel free to ignore!

**Review by dennisameling — APPROVED** [review]
Wow, your PR description is truly outstanding! Thank you so much for picking this up.

This change looks incredibly promising and exciting to me - I wouldn't be surprised if this reduces the runtime of the `setup-git-for-windows-sdk` action from minutes (without cache) to mere seconds. Looking forward to testing it there!!

File: .github/workflows/ci-artifacts.yml (hunk 3 above)
Comment (dscho) [reply .github/workflows/ci-artifacts.yml:231]: This is a once-a-day job, so I'm fine with it running a little slower than absolutely necessary ;-)

File: .github/workflows/ci-artifacts.yml (hunk 2 above)
Comment (dscho) [reply .github/workflows/ci-artifacts.yml]: I did try to add `-mmt=on` (which in my local tests made the `.zip` packing twice as fast). However, both [without](https://github.com/dscho/git-sdk-64/actions/runs/10851317340/job/30114789179#step:11:1) as well as [with](https://github.com/dscho/git-sdk-64/actions/runs/10856380398/job/30130917458#step:11:1) multi-threading, it takes 2 minutes. So I assume that it does not really multi-thread...

Now, we _could_ upload an artifact containing the `7z.exe`, `7.zdll` and `7zCon.sfx` files, then add another job that does the `.zip`/`.7z.exe` packing, and make that another dependency of the `publish-release-assets`  job, but that's just complicating things, and since this workflow is supposed to run at most once a day, I am willing to just be okay with the added 2 minutes. Sound good?

Comment (dennisameling) [reply .github/workflows/ci-artifacts.yml]: > I am willing to just be okay with the added 2 minutes. Sound good?

Yeah, let's leave it as is then. Thanks!
